File: viewer/src/lib/filters.ts
import type { TitleTokenGroup } from "./types.js";

// Returns an array of token groups with exclude flags.
// Single-item group = AND bare word. Multi-item group = OR group (AND-ed with rest).
// Prefix with - to negate: "product -(owner,builder)" → include product, exclude owner/builder
// Input: "product (manager,owner) -(growth,marketing)"
export function parseTitleQuery(input: string): TitleTokenGroup[] {
  const tokens: TitleTokenGroup[] = [];
  const str = input.trim();
  let i = 0;
  while (i < str.length) {
    const isNegated = str[i] === "-";
    if (isNegated) i++;

    if (str[i] === "(") {
      const end = str.indexOf(")", i);
      const inner = end === -1 ? str.slice(i + 1) : str.slice(i + 1, end);
      const terms = inner.split(",").map((t) => t.trim().toLowerCase()).filter(Boolean);
      if (terms.length > 0) tokens.push({ terms, exclude: isNegated });
      i = end === -1 ? str.length : end + 1;
    } else {
      const match = str.slice(i).match(/^[^\s(]+/);
      if (match) {
        const word = match[0].toLowerCase();
        if (word) tokens.push({ terms: [word], exclude: isNegated });
        i += match[0].length;
      } else {
        i++;
      }
    }
  }
  return tokens;
}

// Escape a term for FTS5 query syntax (wrap in double quotes to treat as a phrase).
function ftsEscape(term: string): string {
  return `"${term.replace(/"/g, '""')}"`;
}

// Returns true when all token groups are simple single-term non-negated ANDs —
// the easy case that maps directly to FTS5 AND queries.
function isSimpleFtsQuery(tokens: TitleTokenGroup[]): boolean {
  return tokens.length > 0 && tokens.every((t) => t.terms.length === 1 && !t.exclude);
}

export function addJobFilterConditions(
  filters: {
    title?: string | number | null;
    myLocation?: string | number | null;
    includeRemote?: boolean | null;
    company?: string | number | null;
    sources?: string | string[] | null;
    days?: string | number | null;
    favCompanies?: string[] | null;
    include?: string | null;
    exclude?: string | null;
    tiers?: string | null;
    types?: string | null;
  },
): { conditions: string[]; params: unknown[] } {
  const conditions: string[] = [];
  const params: unknown[] = [];
  const title = String(filters.title ?? "");
  const myLocation = String(filters.myLocation ?? "").trim();
  const includeRemote = filters.includeRemote !== false; // default true
  const company = String(filters.company ?? "");
  const days = String(filters.days ?? "");
  const sources = Array.isArray(filters.sources) ? filters.sources.join(",") : String(filters.sources ?? "");

  if (title.trim()) {
    const tokens = parseTitleQuery(title);

    if (isSimpleFtsQuery(tokens)) {
      // FTS5 path: all plain AND terms — emit a single rowid IN (SELECT rowid FROM fts WHERE title MATCH ?)
      const ftsQuery = tokens.map((t) => ftsEscape(t.terms[0]!)).join(" ");
      conditions.push(
        `rowid IN (SELECT rowid FROM catalog_jobs_fts WHERE catalog_jobs_fts MATCH ?)`,
      );
      params.push(`title:${ftsQuery}`);
    } else {
      // Fallback: complex query (negation / OR groups) — use LIKE as before
      for (const token of tokens) {
        if (token.terms.length === 1) {
          conditions.push(token.exclude ? "LOWER(title) NOT LIKE ?" : "LOWER(title) LIKE ?");
          params.push(`%${token.terms[0]}%`);
        } else {
          if (token.exclude) {
            const notClauses = token.terms.map(() => "LOWER(title) NOT LIKE ?");
            conditions.push(`(${notClauses.join(" AND ")})`);
            for (const t of token.terms) params.push(`%${t}%`);
          } else {
            const orClauses = token.terms.map(() => "LOWER(title) LIKE ?");
            conditions.push(`(${orClauses.join(" OR ")})`);
            for (const t of token.terms) params.push(`%${t}%`);
          }
        }
      }
    }
  }

  // Location: match jobs near myLocation OR remote jobs (if includeRemote).
  // If myLocation is empty and includeRemote is true → remote only.
  // If myLocation is set and includeRemote is false → on-site/hybrid near myLocation only.
  if (myLocation || !includeRemote) {
    const locTerms = myLocation ? myLocation.split(/[,]+/).map((l) => l.trim()).filter(Boolean) : [];
    const locClauses = locTerms.map(() => "LOWER(location) LIKE ?");
    if (includeRemote) {
      // (location matches myLocation) OR is_remote
      const parts = [...locClauses, "is_remote = 1"];
      conditions.push(`(${parts.join(" OR ")})`);
      for (const l of locTerms) params.push(`%${l.toLowerCase()}%`);
    } else {
      // location matches myLocation only (no remote)
      if (locClauses.length > 0) {
        conditions.push(`(${locClauses.join(" OR ")})`);
        for (const l of locTerms) params.push(`%${l.toLowerCase()}%`);
      } else {
        // remote unchecked and no location typed → show nothing (impossible filter, return empty)
        conditions.push("1 = 0");
      }
    }
  }

  if (company.trim()) {
    const companies = company.split(/[,]+/).map((c) => c.trim()).filter(Boolean);
    const companyClauses = companies.map(() => "LOWER(source_key) LIKE ?");
    conditions.push(`(${companyClauses.join(" OR ")})`);
    for (const comp of companies) params.push(`%${comp.toLowerCase()}%`);
  }

  if (sources.trim()) {
    const providerList = sources.split(/[,]+/).map((s) => s.trim()).filter(Boolean);
    if (providerList.length > 0) {
      const providerClauses = providerList.map(() => "provider = ?");
      conditions.push(`(${providerClauses.join(" OR ")})`);
      for (const provider of providerList) params.push(provider);
    }
  }

  if (days.trim()) {
    const n = parseInt(days, 10);
    if (!Number.isNaN(n) && n > 0) {
      conditions.push("COALESCE(posted_at, first_seen_at) >= datetime('now', ?)");
      params.push(`-${n} days`);
    }
  }

  // types: comma-separated canonical employment types to SHOW. NULLs always shown.
  const typeList = String(filters.types ?? "").split(",").map((s) => s.trim()).filter(Boolean);
  if (typeList.length > 0) {
    const typeClauses = typeList.map(() => "employment_type_canonical = ?");
    conditions.push(`(employment_type_canonical IS NULL OR ${typeClauses.join(" OR ")})`);
    for (const t of typeList) params.push(t);
  }

  // tiers: comma-separated tier names to SHOW (intern/entry/mid/senior). If empty = show all.
  // NULLs are included only when all 4 tiers are selected (= no filtering).
  const tierList = String(filters.tiers ?? "").split(",").map((s) => s.trim().toLowerCase()).filter(Boolean);
  if (tierList.length > 0 && tierList.length < 4) {
    const tierClauses = tierList.map(() => "skill_tier = ?");
    conditions.push(`(${tierClauses.join(" OR ")})`);
    for (const tier of tierList) params.push(tier);
  }

  // include: comma-separated terms, each must appear in title OR location OR source_key
  const includeTerms = String(filters.include ?? "").split(",").map((s) => s.trim().toLowerCase()).filter(Boolean);
  for (const term of includeTerms) {
    conditions.push("(LOWER(title) LIKE ? OR LOWER(location) LIKE ? OR LOWER(source_key) LIKE ?)");
    params.push(`%${term}%`, `%${term}%`, `%${term}%`);
  }

  // exclude: comma-separated terms, none may appear in title OR location OR source_key
  const excludeTerms = String(filters.exclude ?? "").split(",").map((s) => s.trim().toLowerCase()).filter(Boolean);
  for (const term of excludeTerms) {
    conditions.push("(LOWER(title) NOT LIKE ? AND LOWER(location) NOT LIKE ? AND LOWER(source_key) NOT LIKE ?)");
    params.push(`%${term}%`, `%${term}%`, `%${term}%`);
  }

  if (filters.favCompanies && filters.favCompanies.length > 0) {
    const favClauses = filters.favCompanies.map(() => "LOWER(source_key) = ?");
    conditions.push(`(${favClauses.join(" OR ")})`);
    for (const fc of filters.favCompanies) params.push(fc.toLowerCase());
  }

  // evaluatedKeys filtering is handled via the analysis_score column in server.ts

  return { conditions, params };
}

File: viewer/backend/services/filters.py
import re

_TOKEN_RE = re.compile(r"[^\s(]+")
_LEADING_INT_RE = re.compile(r"^-?\d+")

ORDER_BY_CLAUSE = "ORDER BY COALESCE(posted_at, first_seen_at) DESC"


def parse_title_query(raw: str) -> list[dict]:
    tokens: list[dict] = []
    s = raw.strip()
    n = len(s)
    i = 0
    while i < n:
        is_negated = s[i] == "-"
        if is_negated:
            i += 1

        if i < n and s[i] == "(":
            end = s.find(")", i)
            inner = s[i + 1 :] if end == -1 else s[i + 1 : end]
            terms = [t.strip().lower() for t in inner.split(",") if t.strip()]
            if terms:
                tokens.append({"terms": terms, "exclude": is_negated})
            i = n if end == -1 else end + 1
        else:
            match = _TOKEN_RE.match(s[i:])
            if match:
                word = match.group(0).lower()
                if word:
                    tokens.append({"terms": [word], "exclude": is_negated})
                i += len(match.group(0))
            else:
                i += 1
    return tokens


def _fts_escape(term: str) -> str:
    return '"' + term.replace('"', '""') + '"'


def _is_simple_fts_query(tokens: list[dict]) -> bool:
    return len(tokens) > 0 and all(
        len(t["terms"]) == 1 and not t["exclude"] for t in tokens
    )


def _parse_int(raw: str) -> int | None:
    match = _LEADING_INT_RE.match(raw.strip())
    return int(match.group(0)) if match else None


def add_job_filter_conditions(
    title: str | None = None,
    my_location: str | None = None,
    include_remote: bool = True,
    company: str | None = None,
    sources: str | list[str] | None = None,
    days: str | int | None = None,
    fav_companies: list[str] | None = None,
    include: str | None = None,
    exclude: str | None = None,
    tiers: str | None = None,
    types: str | None = None,
    evaluated: bool = False,
    score: str | None = None,
) -> tuple[str, list]:
    conditions: list[str] = []
    params: list = []

    title = str(title) if title is not None else ""
    my_location = str(my_location).strip() if my_location is not None else ""
    company = str(company) if company is not None else ""
    days_str = str(days) if days is not None else ""
    sources_str = (
        ",".join(sources)
        if isinstance(sources, list)
        else str(sources) if sources is not None else ""
    )

    if title.strip():
        tokens = parse_title_query(title)

        if _is_simple_fts_query(tokens):
            fts_query = " ".join(_fts_escape(t["terms"][0]) for t in tokens)
            conditions.append(
                "rowid IN (SELECT rowid FROM catalog_jobs_fts WHERE catalog_jobs_fts MATCH ?)"
            )
            params.append(f"title:{fts_query}")
        else:
            for token in tokens:
                terms = token["terms"]
                if len(terms) == 1:
                    conditions.append(
                        "LOWER(title) NOT LIKE ?"
                        if token["exclude"]
                        else "LOWER(title) LIKE ?"
                    )
                    params.append(f"%{terms[0]}%")
                elif token["exclude"]:
                    conditions.append(
                        "("
                        + " AND ".join("LOWER(title) NOT LIKE ?" for _ in terms)
                        + ")"
                    )
                    params.extend(f"%{t}%" for t in terms)
                else:
                    conditions.append(
                        "(" + " OR ".join("LOWER(title) LIKE ?" for _ in terms) + ")"
                    )
                    params.extend(f"%{t}%" for t in terms)

    # Location: match jobs near my_location OR remote jobs (if include_remote).
    # Empty my_location + include_remote -> remote only.
    # my_location set + include_remote False -> on-site/hybrid near my_location only.
    if my_location or not include_remote:
        loc_terms = (
            [l.strip() for l in my_location.split(",") if l.strip()]
            if my_location
            else []
        )
        loc_clauses = ["LOWER(location) LIKE ?" for _ in loc_terms]
        if include_remote:
            conditions.append("(" + " OR ".join([*loc_clauses, "is_remote = 1"]) + ")")
            params.extend(f"%{l.lower()}%" for l in loc_terms)
        elif loc_clauses:
            conditions.append("(" + " OR ".join(loc_clauses) + ")")
            params.extend(f"%{l.lower()}%" for l in loc_terms)
        else:
            # remote excluded and no location given -> impossible filter, return nothing
            conditions.append("1 = 0")

    if company.strip():
        companies = [c.strip() for c in company.split(",") if c.strip()]
        conditions.append(
            "(" + " OR ".join("LOWER(source_key) LIKE ?" for _ in companies) + ")"
        )
        params.extend(f"%{c.lower()}%" for c in companies)

    if sources_str.strip():
        provider_list = [s.strip() for s in sources_str.split(",") if s.strip()]
        if provider_list:
            conditions.append(
                "(" + " OR ".join("provider = ?" for _ in provider_list) + ")"
            )
            params.extend(provider_list)

    if days_str.strip():
        n = _parse_int(days_str)
        if n is not None and n > 0:
            conditions.append(
                "COALESCE(posted_at, first_seen_at) >= datetime('now', ?)"
            )
            params.append(f"-{n} days")

    # types: canonical employment types to show. NULLs always included.
    type_list = [t.strip() for t in str(types or "").split(",") if t.strip()]
    if type_list:
        type_clauses = " OR ".join("employment_type_canonical = ?" for _ in type_list)
        conditions.append(f"(employment_type_canonical IS NULL OR {type_clauses})")
        params.extend(type_list)

    # tiers: skill_tier values to show. Selecting all 4 == no filter (NULLs shown too).
    tier_list = [t.strip().lower() for t in str(tiers or "").split(",") if t.strip()]
    if 0 < len(tier_list) < 4:
        conditions.append("(" + " OR ".join("skill_tier = ?" for _ in tier_list) + ")")
        params.extend(tier_list)

    # include: each term must appear in title OR location OR source_key
    include_terms = [
        t.strip().lower() for t in str(include or "").split(",") if t.strip()
    ]
    for term in include_terms:
        conditions.append(
            "(LOWER(title) LIKE ? OR LOWER(location) LIKE ? OR LOWER(source_key) LIKE ?)"
        )
        params.extend([f"%{term}%", f"%{term}%", f"%{term}%"])

    # exclude: no term may appear in title OR location OR source_key
    exclude_terms = [
        t.strip().lower() for t in str(exclude or "").split(",") if t.strip()
    ]
    for term in exclude_terms:
        conditions.append(
            "(LOWER(title) NOT LIKE ? AND LOWER(location) NOT LIKE ? AND LOWER(source_key) NOT LIKE ?)"
        )
        params.extend([f"%{term}%", f"%{term}%", f"%{term}%"])

    if fav_companies:
        conditions.append(
            "(" + " OR ".join("LOWER(source_key) = ?" for _ in fav_companies) + ")"
        )
        params.extend(fc.lower() for fc in fav_companies)

    if evaluated:
        conditions.append("analysis_score IS NOT NULL AND analysis_score > 0")
    if score == "none":
        conditions.append("(analysis_score IS NULL OR analysis_score = 0)")
    elif score == "4plus":
        conditions.append("analysis_score >= 4")

    where_clause = f"WHERE {' AND '.join(conditions)}" if conditions else ""
    return where_clause, params
